Allow apps to pass along own user agent

mdarby opened the pull request:
In relation to: https://github.com/fog/fog/issues/3433#issuecomment-73893922

geemus commented on lib/fog/core/connection.rb:
```diff
@@ -77,6 +75,13 @@ def reset
       private
 
+      def user_agent
+        arr = [ENV['FOG_CUSTOM_USER_AGENT']] # Allow apps to pass custom UA for metrics, etc.
```
This should probably also use an `if` instead of just adding nil (otherwise when you join you end up with leading whitespace, which is why the tests break).

geemus replied:
and/or lstrip or something.

mdarby replied:
I added `compact`down here to take care of that.

everett-toews commented:
Can this be documented in the source somewhere that users can become aware of it?

Any such documentation should also include an example of how to set a proper User-Agent as it's really easy to get wrong. e.g. `my-app/1.2`

mdarby commented:
Oh no doubts -- this was just to make sure we were on the same page.

fcheung commented:
Any reason to prefer an env variable over an option passed to `initialize` ?

mdarby commented:
Ease honestly.

This connection method is really quite buried and I wouldn't want to manually pass a bit of info through the layers of Fog. Seems to me that if someone wanted to use this feature they could more easily control it via an optional ENV var.

fcheung commented:
How about a class level property `Fog::Connection.user_agent = ...` ?

mdarby commented:
If you'd prefer that method I can easily refactor. :)

plribeiro3000 commented:
I think that perhaps a variable through fog would better because the user would be able to use both options. He can define the code to search on `ENV` manually and thus have the same behavior. But thats just my opinion. =)

fcheung commented:
I feel a bit squiffy about using environment variables, unless the intention truly is that one should be able to change this without changing source code

tokengeek commented:
I think I'm with @fcheung on this. `ENV` are great for applications but `fog-core` and the majority of `fog` is intended for use as a library.

If people want to see that `Carrierwave` is making requests, `carrierwave` shouldn't be injecting ENVs into other people's applications.

That's my take on it anyway.

The majority of the `ENV` settings in `fog` are for using the binary but maybe some leaked through.

mdarby commented:
Thoughts?

![1 irb ruby 2015-02-12 12-07-43](https://cloud.githubusercontent.com/assets/2805/6172675/3bf5823a-b2b0-11e4-8328-a9df5bb8b29c.jpg)

fcheung commented:
I think that if you do it like that subclasses (e.g. Fog::XML::Connection) won't inherit the value. Perhaps one of the few times where an @@var is appropriate

mdarby commented:
I've refactored to use `@@var`.
I added a `Fog::Connection.add_user_agent(str)` method that checks for format.

![updated](https://cloud.githubusercontent.com/assets/2805/6173861/1134069e-b2b9-11e4-89cd-38b8c4c13d9e.jpg)

mdarby commented:
Direction approval?

geemus commented:
Seems reasonable to me, I'll let the others add their 2 cents though. Also, I'm not altogether opposed to ENV also overriding this, I just agree that I would prefer for it to not be the ONLY way of setting this value. I think there can be some value to supporting both.

mdarby commented:
Bump

GTG for docs/tests?

plribeiro3000 commented:
`Fog::Core::Connection.add_user_agent("my-app/1.2")` seem reasonable.

:+1: from me!

geemus commented:
Thanks!

geemus commented:
released in 1.29.0
